Added triton interop support (#36)
Huawei-CPLLab/PyDSL: docs/usage.md

The argument to the autotune is an object of type `AutotuneConfigurations`, which is a very thin wrapper around a list of `Config`. Each config in the list is compiled/ran and performance is measured. After this is complete, the function is set to the fastest one measured.

Note that `AutotuneConfigurations` provides a number of utility functions to help users build lists of configurations. These include primitives to do cartesian products, zipping and concatenation. These make it easy to produce comprehensive search spaces.

# Inter-operability with Triton

PyDSL supports inter-operability with Triton, i.e. it is possible to call a Triton function from within a PyDSL function. This works by first compiling the triton function to triton MLIR (ttir), converting it to the linalg dialect using the triton-adapter-opt tool, and then combining it with the MLIR generated by PyDSL.

## Requirements

> in order to use the inter-op feature of PyDSL, you require a configured triton environment, as well as the tool triton-adapter-opt from [triton-ascend](https://gitee.com/ascend/triton-ascend). 

Reviewer: ```suggestion
in order to use the inter-op feature of PyDSL, you require a configured Triton environment, as well as the tool triton-adapter-opt from [triton-ascend](https://gitee.com/ascend/triton-ascend). 
```

> ### Building triton-ascend and adding triton-adapter-opt to your PATH:
> 
> #### using pip:
> 
> For people who don't want to build all of triton-ascend, it is possible to extract only the triton-adapter-opt binary using the python wheel for triton-ascend. Since this is designed for python version 3.11 and we support python3.12, you can use the following commands to extract the triton-adapter-opt binary, add it to your PATH, and then remove the wheel. Make sure to execute this command wherever you want to have the triton-adapter-opt binary stored (such as the PyDSL root directory):

Reviewer: ```suggestion
### Building triton-ascend and adding triton-adapter-opt to your PATH

#### Using pip

For people who don't want to build all of triton-ascend, it is possible to extract only the triton-adapter-opt binary using the Python wheel for triton-ascend. Since this is designed for Python version 3.11 and we support Python 3.12, you can use the following commands to extract the triton-adapter-opt binary, add it to your PATH, and then remove the wheel. Make sure to execute this command wherever you want to have the triton-adapter-opt binary stored (such as the PyDSL root directory):
```

```bash
pip download triton-ascend --python-version=3.11 --no-deps
unzip $(find -name "triton_ascend*.whl") -d extracted_wheel
mkdir triton_bin
cp extracted_wheel/triton/backends/huawei/triton-adapter-opt triton_bin
rm -rf extracted_wheel
rm -rf $(find -name "triton_ascend*.whl")
export PATH=$(pwd)/triton_bin:$PATH
```

> #### building from source:
> 
> build instructions for building triton-ascend can be found here:  https://gitee.com/ascend/triton-ascend.
> After building, you need to add the triton-adapter-opt tool to you path, i.e.
> ```bash
> export PATH=$(find $ASCEND_DIRECTORY/build -name "triton-adapter-opt" | head -n 1):$PATH
> ```
> changing $ASCEND_DIRECTORY to point to the location of your triton-ascend directory.

Reviewer: ```suggestion
#### Building from source

Build instructions for building triton-ascend can be found here: https://gitee.com/ascend/triton-ascend.
After building, you need to add the triton-adapter-opt tool to you path, i.e.
```bash
export PATH=$(find $ASCEND_DIRECTORY/build -name "triton-adapter-opt" | head -n 1):$PATH
```
changing `$ASCEND_DIRECTORY` to point to the location of your triton-ascend directory.
```

## Examples/Usage

The basic usage is to simply call an `@triton.jit` function from within a PyDSL function:

> ```python
> @triton.jit
> def kernel(a_ptr, ...): # signature
>     pid_x = tl.program_id(axis=0)
>     pid_y = tl.program_id(axis=1)
>     pid_z = tl.program_id(axis=2)
> 	# triton implmentation goes here
> 
> @compile()
> def pydsl_func():
> 	# pydsl code goes here
> 	for i in arange(x_size):
>         for j in arange(y_size):
>             for k in arange(z_size):
> 	            kernel(a_ptr, ..., i, j, k) 
> 	# more pydsl code
> ```

Reviewer: ```suggestion
```python
@triton.jit
def kernel(a_ptr, ...): # signature
    pid_x = tl.program_id(axis=0)
    pid_y = tl.program_id(axis=1)
    pid_z = tl.program_id(axis=2)
    # triton implmentation goes here

@compile()
def pydsl_func():
    # pydsl code goes here
    for i in arange(x_size):
        for j in arange(y_size):
            for k in arange(z_size):
	            kernel(a_ptr, ..., i, j, k) 
    # more pydsl code
```
```

> Note how in addition to the parameters defined by the kernel, you can also provide 3 extra parameters, which defines the grid index that you want to call the kernel with (x, y, z). These are `Index` arguments which need to be added for each axis that is defined within the triton kernel. A more detailed example of how to use the interop feature can be found in [examples/triton_interop.py](/examples/triton_interop.py).

Reviewer: ```suggestion
Note how in addition to the parameters defined by the kernel, you can also provide 3 extra parameters (x, y, z), which define the grid index that you want to call the kernel with. These are `Index` arguments which need to be added for each axis that is defined within the Triton kernel. A more detailed example of how to use the interop feature can be found in [examples/triton_interop.py](/examples/triton_interop.py).
```

For calling a triton function with a constexpr parameter, a `Number` object that has an explicit name must be passed in, i.e.
```python
@triton.jit
def kernel(... a: tl.constexpr, ...):
	# triton implementation goes here

@compile()
def pydsl_func():
	# pydsl code
	a: Number = 64
	kernel(..., a, ...)
```

## Limitations

> - Since this feature uses triton-adapter-opt to convert the triton dialect to the linalg dialect, it inherets the limitations of triton ascend. This includes the fact that not all ops are currently supported for all data types, and uint is currently not supported at all. Current support is detailed [here]([docs/sources/python-api/outline.md · Ascend/triton-ascend - Gitee](https://gitee.com/ascend/triton-ascend/blob/master/docs/sources/python-api/outline.md)) 
> 
> - PyDSL `Tensor` types are currently not supported, and all arrays passed in must be `Memrefs`.
> 
> - `Memref` shapes must always be 1-dimensional `DYNAMIC` arrays when they are passed into a triton function.
> 
> - triton-adapter-opt only supports lowering a single function. Triton inlines function calls by default,
> but if a function is called that has noinline=true then the compilation will fail

Reviewer: ```suggestion
- Since this feature uses triton-adapter-opt to convert the triton dialect to the linalg dialect, it inherits the limitations of Triton Ascend. This includes the fact that not all ops are currently supported for all data types, and uint is currently not supported at all. Current support is detailed [here]([docs/sources/python-api/outline.md · Ascend/triton-ascend - Gitee](https://gitee.com/ascend/triton-ascend/blob/master/docs/sources/python-api/outline.md)) 

- PyDSL `Tensor` types are currently not supported; all arrays passed in must be `MemRef`s.

- `MemRef` shapes must always be 1-dimensional `DYNAMIC` arrays when they are passed into a Triton function.

- triton-adapter-opt only supports lowering a single function. Triton inlines function calls by default,
but if a function is called that has `noinline=true` then the compilation will fail
```

Also we should have a testcase to check that having a `noinline=true` function indeed throws a reasonable error and doesn't just blow up. (Use `compilation_failed_from` in `tests/helper.py`).

Author: I agree that that would be a good test case, but if you try to define two `triton.jit` functions within another function and have one call the other, they won't be able to. They need to be in global scope to be called, so they would need to be defined outside of the test which isn't ideal

Author: also if you do try to run a noinline function you get a "CalledProcessError: Command 'triton-adapter-opt --triton-to-linalg' returned non-zero exit status 1".
